rpc.go

```diff
@@ -293,11 +303,106 @@ func (l *Libvirt) deregister(id uint32) {
 // returns response returned by server.
 // if response is not OK, decodes error from it and returns it.
 func (l *Libvirt) request(proc uint32, program uint32, payload []byte) (response, error) {
+	return l.requestStream(proc, program, payload, nil, nil)
+}
+
+func (l *Libvirt) requestStream(proc uint32, program uint32, payload []byte, outStream io.Reader, inStream io.Writer) (response, error) {
 	serial := l.serial()
 	c := make(chan response)
 
 	l.register(serial, c)
+	defer l.deregister(serial)
+
+	err := l.sendPacket(serial, proc, program, payload, Call, StatusOK)
+	if err != nil {
+		return response{}, err
+	}
+
+	resp, err := l.getResponse(c)
+	if err != nil {
+		return resp, err
+	}
+
+	var abortOutStream chan bool
+	var outStreamErr chan error
+	if outStream != nil {
+		abortOutStream = make(chan bool)
+		outStreamErr = make(chan error)
+		go func() {
+			var err error
+			var n int
+			buf := make([]byte, 1<<22-24)
+			for {
+				select {
+				case <-abortOutStream:
+					err = l.sendPacket(serial, proc, program, nil, Stream, StatusError)
+					break
+				default:
+				}
+				n, err = outStream.Read(buf)
+				if err != nil {
+					if err == io.EOF {
```

Comment:
return as early as possible here
```go
outstreamErr <- l.sendPacket(serial, proc.program, nil, Stream, StatusOK)
return
```

Reply:
I wanted to avoid repeating this snippet all over the place. I'll factor out this peace into separate function.